### DISCUSSION TITLE
Adding tables to scenario bundles via API

### DISCUSSION THREAD

**Discussion by ChristophSchmidtinRET** [openego/oeplatform#1927] [Q&A] [answered]
Hello everyone,

when uploading a table to the oep, is it possible to add the table to a scenario bundle? Or do I have to do this manually afterwards?

Best regards

**Comment by jh-RLI** [comment]
Hey, this is work in progress. Once this linked Pull Request is merged. Should have time to finalize it in January. 
https://github.com/OpenEnergyPlatform/oeplatform/pull/1894

The feature will be similar to what you describe. I can add it to the table upload step, but it is designed to be used for any table at any time if the table is uploaded to the OEP and published in the "scenario" topic. I hope to also have some spare time to add a User interface to do this, but at least there will be a detailed description on how to use the feature via the WEB-API (using http requests). 

The Scenario Bundle UI will also be updated to enable selecting multiple tables at once. This might take more time.

**Comment by jh-RLI** [comment]
Hello again, sorry for the long delay. Next week, I will finally have some time to finalize the pull request. After that I can make a new release and install everything. After that it would be great if we could meet and I can instruct you how to work with the new API endpoint and check if the documentation is understandable.

**Reply by jh-RLI** [reply]
Im still working on the PR to implement the new API ... as i tried out a better approach to send SparqlQueries (should make them faster) the changes are quite substantial. Currently, i have to do some more testing and fine-tune the Query to insert the information in the correct structure to the OEKG. Currently it seems to me that i can finish it this week. After that there will be an OEP release and i will install the code updates on openenergyplatform.org

**Comment by jh-RLI** [comment]
This is off topic but related to the scenario bundles: 
The feature to enable usergroup based editing of scneario bundles will also be added to soon. There will be access based on the API-Token (if the user is part of a group with access). Im planning on using the same groups that are used for table resource access management. I hope this will make things qick and easy. Im also planning on working on this during this week.

**Comment by jh-RLI** [comment]
Another user - request mentioned that it would be great to be able to link datasets (files in any form and at any location) that are stored outside the oeplatform. I will try to add this possibility while implementing the API. It will be an API only feature for now, i got no time to adjust the UI.  Other services that we use: namely the https://databus.openenergyplatform.org/ already enables registering datasets from any storage location. So this requirement fits in our framework anyway.

But it should be noted that the current implementation of the Scenario bundles is coupled to comparison aspects (qualitative & quantitative comparison of scenario projection data). Especially the quantitative comparison is only possible once the data itself is accessible via the OEP since its backend requires having access to a SQL database. 
Still this functionality is not yet complete and therefore hard to use for "any" user. It also requires more steps like adding ontological annotations that describe the structure of the data, and it can only be applied to tabular data resources.

**Comment by jh-RLI** [comment]
In https://github.com/OpenEnergyPlatform/oeplatform/pull/1894/ adding datasets using the API is now implemented. I will install it soon.  External datasets can be added but they must be registered on https://databus.openenergyplatform.org/ to provide persistent id´s .
I created a basic documentation page which i will link once available. 

Deleting datasets using the API is not yet implemented.

**Comment by jh-RLI** [comment, marked as answer]
https://openenergyplatform.github.io/oeplatform/oeplatform-code/web-api/oekg-api/scenario-dataset/
